src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -184,9 +130,541 @@ def _validate_certificate_with_key(certificate, private_key, schema_name, verror
+        'cryptokey.validate_certificate_with_key', certificate, private_key, schema_name, verrors, passphrase
+    )
+
+    key_type = data.get('key_type')
+    if key_type:
+        if key_type != 'EC' and not data.get('key_length'):
+            verrors.add(
+                f'{schema_name}.key_length',
+                'This field is required with RSA based keys'
+            )
+
+
+class CryptoKeyService(Service):
+
+    ec_curve_default = 'BrainpoolP384R1'
+
+    ec_curves = [
+        'BrainpoolP512R1',
+        'BrainpoolP384R1',
+        'BrainpoolP256R1',
+        'SECP256K1'
+    ]
+
+    backend_mappings = {
+        'common_name': 'common',
+        'country_name': 'country',
+        'state_or_province_name': 'state',
+        'locality_name': 'city',
+        'organization_name': 'organization',
+        'organizational_unit_name': 'organizational_unit',
+        'email_address': 'email'
+    }
+
+    class Config:
+        private = True
+
+    def validate_certificate_with_key(self, certificate, private_key, schema_name, verrors, passphrase=None):
+        if (
+            (certificate and private_key) and
+            all(k not in verrors for k in (f'{schema_name}.certificate', f'{schema_name}.privatekey'))
+        ):
+            public_key_obj = crypto.load_certificate(crypto.FILETYPE_PEM, certificate)
+            private_key_obj = crypto.load_privatekey(
+                crypto.FILETYPE_PEM,
+                private_key,
+                passphrase=passphrase.encode() if passphrase else None
+            )
+
+            try:
+                context = SSL.Context(SSL.TLSv1_2_METHOD)
+                context.use_certificate(public_key_obj)
+                context.use_privatekey(private_key_obj)
+                context.check_privatekey()
+            except SSL.Error as e:
+                verrors.add(
+                    f'{schema_name}.privatekey',
+                    f'Private key does not match certificate: {e}'
+                )
+
+        return verrors
+
+    def validate_private_key(self, private_key, verrors, schema_name, passphrase=None):
+        private_key_obj = self.load_private_key(private_key, passphrase)
+        if not private_key_obj:
+            verrors.add(
+                f'{schema_name}.privatekey',
+                'Please provide a valid private key with matching passphrase ( if any )'
+            )
+        elif (
+            'create' in schema_name and private_key_obj.key_size < 1024 and not isinstance(
+                private_key_obj, ec.EllipticCurvePrivateKey
+            )
+        ):
+            # When a cert/ca is being created, we disallow keys with size less then 1024
+            # Update is allowed for now for keeping compatibility with very old cert/keys
+            # We do not do this check for any EC based key
+            verrors.add(
+                f'{schema_name}.privatekey',
+                'Please provide a key with size greater than or equal to 1024 bits'
```

Comment:
Again, "please" implies it's not a requirement.  Instead:
```suggestion
                'Key size must be greater than or equal to 1024 bits.'
```